{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "adb7f1a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"1002\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  const force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  const JS_MIME_TYPE = 'application/javascript';\n",
       "  const HTML_MIME_TYPE = 'text/html';\n",
       "  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  const CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    const script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    const cell = handle.cell;\n",
       "\n",
       "    const id = cell.output_area._bokeh_element_id;\n",
       "    const server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd_clean, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            const id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd_destroy);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    const output_area = handle.output_area;\n",
       "    const output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      const bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      const script_attrs = bk_div.children[0].attributes;\n",
       "      for (let i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      const toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    const events = require('base/js/events');\n",
       "    const OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  const NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    const el = document.getElementById(\"1002\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  \n",
       "  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.1.min.js\"];\n",
       "  const css_urls = [];\n",
       "  \n",
       "\n",
       "  const inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    function(Bokeh) {\n",
       "    \n",
       "    \n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "      \n",
       "    for (let i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "    if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      const cell = $(document.getElementById(\"1002\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(\"1002\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  \n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.1.min.js\"];\n  const css_urls = [];\n  \n\n  const inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    function(Bokeh) {\n    \n    \n    }\n  ];\n\n  function run_inline_js() {\n    \n    if (root.Bokeh !== undefined || force === true) {\n      \n    for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n    if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(\"1002\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from __future__ import print_function\n",
    "import re, gzip, time, itertools, io\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import requests\n",
    "import matplotlib.pyplot as plt\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "import pandas_bokeh\n",
    "from bokeh.io import output_file, show\n",
    "pandas_bokeh.output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fec68e05",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('Homo_sapiens.GRCh38.pep.all.fa') as fasta_file:  # Will close handle cleanly\n",
    "    identifiers = []\n",
    "    lengths = []\n",
    "    sequences = []\n",
    "    annotation =[]\n",
    "    for title, sequence in SimpleFastaParser(fasta_file):\n",
    "        identifiers.append(title.split(None, 1)[0]) # First word is ID\n",
    "        annotation.append(title)\n",
    "        lengths.append(len(sequence))\n",
    "        sequences.append(sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "40870493",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 116646 entries, 0 to 116645\n",
      "Data columns (total 4 columns):\n",
      " #   Column      Non-Null Count   Dtype \n",
      "---  ------      --------------   ----- \n",
      " 0   ID          116646 non-null  object\n",
      " 1   seq_length  116646 non-null  int64 \n",
      " 2   sequence    116646 non-null  object\n",
      " 3   annotation  116646 non-null  object\n",
      "dtypes: int64(1), object(3)\n",
      "memory usage: 3.6+ MB\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>seq_length</th>\n",
       "      <th>sequence</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSP00000488240.1</td>\n",
       "      <td>4</td>\n",
       "      <td>GTGG</td>\n",
       "      <td>ENSP00000488240.1 pep chromosome:GRCh38:CHR_HS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSP00000451042.1</td>\n",
       "      <td>2</td>\n",
       "      <td>EI</td>\n",
       "      <td>ENSP00000451042.1 pep chromosome:GRCh38:14:224...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSP00000452494.1</td>\n",
       "      <td>4</td>\n",
       "      <td>TGGY</td>\n",
       "      <td>ENSP00000452494.1 pep chromosome:GRCh38:14:224...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSP00000451515.1</td>\n",
       "      <td>3</td>\n",
       "      <td>PSY</td>\n",
       "      <td>ENSP00000451515.1 pep chromosome:GRCh38:14:224...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSP00000487941.1</td>\n",
       "      <td>4</td>\n",
       "      <td>GTGG</td>\n",
       "      <td>ENSP00000487941.1 pep chromosome:GRCh38:7:1427...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  ID  seq_length sequence  \\\n",
       "0  ENSP00000488240.1           4     GTGG   \n",
       "1  ENSP00000451042.1           2       EI   \n",
       "2  ENSP00000452494.1           4     TGGY   \n",
       "3  ENSP00000451515.1           3      PSY   \n",
       "4  ENSP00000487941.1           4     GTGG   \n",
       "\n",
       "                                          annotation  \n",
       "0  ENSP00000488240.1 pep chromosome:GRCh38:CHR_HS...  \n",
       "1  ENSP00000451042.1 pep chromosome:GRCh38:14:224...  \n",
       "2  ENSP00000452494.1 pep chromosome:GRCh38:14:224...  \n",
       "3  ENSP00000451515.1 pep chromosome:GRCh38:14:224...  \n",
       "4  ENSP00000487941.1 pep chromosome:GRCh38:7:1427...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#converting lists to pandas Series    \n",
    "s1 = pd.Series(identifiers, name='ID')\n",
    "s2 = pd.Series(lengths, name='seq_length')\n",
    "s3 = pd.Series(sequences, name='sequence')\n",
    "s4 = pd.Series(annotation, name='annotation')\n",
    "#Gathering Series into a pandas DataFrame and rename index as ID column\n",
    "df = pd.DataFrame(dict(ID=s1, seq_length=s2,sequence=s3, annotation=s4 ))\n",
    "df.info()\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "25e11d37",
   "metadata": {},
   "outputs": [],
   "source": [
    "p = re.compile(r'(?P<pep_ID>[^\\s]+)\\s+'\n",
    "                   + r'(?P<seqtype>[^:\\s]+):?.'\n",
    "                   + r'chromosome:(?P<version>[^:]+):'\n",
    "                   + r'(?P<chromosome>[^:]+):'\n",
    "                   + r'(?P<start>[^:]+):'\n",
    "                   + r'(?P<end>[^:]+):'\n",
    "                   + r'(?P<strand>[^\\s]+)\\s+'\n",
    "                   + r'gene:(?P<gene_ID>[^\\s]+)\\s+'\n",
    "                   + r'transcript:(?P<transcript_ID>[^\\s]+)\\s*'\n",
    "                   + r'gene_biotype:(?P<gene_biotype>[^\\s]+)\\s*'\n",
    "                   + r'transcript_biotype:(?P<transcript_biotype>[^\\s]+)\\s*'\n",
    "                   + r'gene_symbol:(?P<gene_symbol>[^\\s]+)\\s*'\n",
    "                   + r'description:(?P<description>[^[]+)\\s'\n",
    "                   + r'\\[Source:(?P<source>[^;]+);*'\n",
    "                   + r'Acc:(?P<accession>[^]]+)\\]*')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "88a5962f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>seq_length</th>\n",
       "      <th>sequence</th>\n",
       "      <th>annotation</th>\n",
       "      <th>pep_ID</th>\n",
       "      <th>seqtype</th>\n",
       "      <th>version</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>transcript_biotype</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>description</th>\n",
       "      <th>source</th>\n",
       "      <th>accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSP00000488240.1</td>\n",
       "      <td>4</td>\n",
       "      <td>GTGG</td>\n",
       "      <td>ENSP00000488240.1 pep chromosome:GRCh38:CHR_HS...</td>\n",
       "      <td>ENSP00000488240.1</td>\n",
       "      <td>pep</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>CHR_HSCHR7_2_CTG6</td>\n",
       "      <td>142847306</td>\n",
       "      <td>142847317</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000282253.1</td>\n",
       "      <td>ENST00000631435.1</td>\n",
       "      <td>TR_D_gene</td>\n",
       "      <td>TR_D_gene</td>\n",
       "      <td>TRBD1</td>\n",
       "      <td>T cell receptor beta diversity 1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>HGNC:12158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSP00000451042.1</td>\n",
       "      <td>2</td>\n",
       "      <td>EI</td>\n",
       "      <td>ENSP00000451042.1 pep chromosome:GRCh38:14:224...</td>\n",
       "      <td>ENSP00000451042.1</td>\n",
       "      <td>pep</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>14</td>\n",
       "      <td>22438547</td>\n",
       "      <td>22438554</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000223997.1</td>\n",
       "      <td>ENST00000415118.1</td>\n",
       "      <td>TR_D_gene</td>\n",
       "      <td>TR_D_gene</td>\n",
       "      <td>TRDD1</td>\n",
       "      <td>T cell receptor delta diversity 1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>HGNC:12254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSP00000452494.1</td>\n",
       "      <td>4</td>\n",
       "      <td>TGGY</td>\n",
       "      <td>ENSP00000452494.1 pep chromosome:GRCh38:14:224...</td>\n",
       "      <td>ENSP00000452494.1</td>\n",
       "      <td>pep</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>14</td>\n",
       "      <td>22449113</td>\n",
       "      <td>22449125</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000228985.1</td>\n",
       "      <td>ENST00000448914.1</td>\n",
       "      <td>TR_D_gene</td>\n",
       "      <td>TR_D_gene</td>\n",
       "      <td>TRDD3</td>\n",
       "      <td>T cell receptor delta diversity 3</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>HGNC:12256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSP00000451515.1</td>\n",
       "      <td>3</td>\n",
       "      <td>PSY</td>\n",
       "      <td>ENSP00000451515.1 pep chromosome:GRCh38:14:224...</td>\n",
       "      <td>ENSP00000451515.1</td>\n",
       "      <td>pep</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>14</td>\n",
       "      <td>22439007</td>\n",
       "      <td>22439015</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000237235.2</td>\n",
       "      <td>ENST00000434970.2</td>\n",
       "      <td>TR_D_gene</td>\n",
       "      <td>TR_D_gene</td>\n",
       "      <td>TRDD2</td>\n",
       "      <td>T cell receptor delta diversity 2</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>HGNC:12255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSP00000487941.1</td>\n",
       "      <td>4</td>\n",
       "      <td>GTGG</td>\n",
       "      <td>ENSP00000487941.1 pep chromosome:GRCh38:7:1427...</td>\n",
       "      <td>ENSP00000487941.1</td>\n",
       "      <td>pep</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>7</td>\n",
       "      <td>142786213</td>\n",
       "      <td>142786224</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000282431.1</td>\n",
       "      <td>ENST00000632684.1</td>\n",
       "      <td>TR_D_gene</td>\n",
       "      <td>TR_D_gene</td>\n",
       "      <td>TRBD1</td>\n",
       "      <td>T cell receptor beta diversity 1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>HGNC:12158</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  ID  seq_length sequence  \\\n",
       "0  ENSP00000488240.1           4     GTGG   \n",
       "1  ENSP00000451042.1           2       EI   \n",
       "2  ENSP00000452494.1           4     TGGY   \n",
       "3  ENSP00000451515.1           3      PSY   \n",
       "4  ENSP00000487941.1           4     GTGG   \n",
       "\n",
       "                                          annotation             pep_ID  \\\n",
       "0  ENSP00000488240.1 pep chromosome:GRCh38:CHR_HS...  ENSP00000488240.1   \n",
       "1  ENSP00000451042.1 pep chromosome:GRCh38:14:224...  ENSP00000451042.1   \n",
       "2  ENSP00000452494.1 pep chromosome:GRCh38:14:224...  ENSP00000452494.1   \n",
       "3  ENSP00000451515.1 pep chromosome:GRCh38:14:224...  ENSP00000451515.1   \n",
       "4  ENSP00000487941.1 pep chromosome:GRCh38:7:1427...  ENSP00000487941.1   \n",
       "\n",
       "  seqtype version         chromosome      start        end strand  \\\n",
       "0     pep  GRCh38  CHR_HSCHR7_2_CTG6  142847306  142847317      1   \n",
       "1     pep  GRCh38                 14   22438547   22438554      1   \n",
       "2     pep  GRCh38                 14   22449113   22449125      1   \n",
       "3     pep  GRCh38                 14   22439007   22439015      1   \n",
       "4     pep  GRCh38                  7  142786213  142786224      1   \n",
       "\n",
       "             gene_ID      transcript_ID gene_biotype transcript_biotype  \\\n",
       "0  ENSG00000282253.1  ENST00000631435.1    TR_D_gene          TR_D_gene   \n",
       "1  ENSG00000223997.1  ENST00000415118.1    TR_D_gene          TR_D_gene   \n",
       "2  ENSG00000228985.1  ENST00000448914.1    TR_D_gene          TR_D_gene   \n",
       "3  ENSG00000237235.2  ENST00000434970.2    TR_D_gene          TR_D_gene   \n",
       "4  ENSG00000282431.1  ENST00000632684.1    TR_D_gene          TR_D_gene   \n",
       "\n",
       "  gene_symbol                        description       source   accession  \n",
       "0       TRBD1   T cell receptor beta diversity 1  HGNC Symbol  HGNC:12158  \n",
       "1       TRDD1  T cell receptor delta diversity 1  HGNC Symbol  HGNC:12254  \n",
       "2       TRDD3  T cell receptor delta diversity 3  HGNC Symbol  HGNC:12256  \n",
       "3       TRDD2  T cell receptor delta diversity 2  HGNC Symbol  HGNC:12255  \n",
       "4       TRBD1   T cell receptor beta diversity 1  HGNC Symbol  HGNC:12158  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.concat([\n",
    "    df,\n",
    "    (\n",
    "        df.annotation.str.extractall(p)\n",
    "          .reset_index('match', drop=True)\n",
    "    )\n",
    "], axis=1).fillna('')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "92713561",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 95092 entries, 762 to 116645\n",
      "Data columns (total 18 columns):\n",
      " #   Column              Non-Null Count  Dtype \n",
      "---  ------              --------------  ----- \n",
      " 0   ID                  95092 non-null  object\n",
      " 1   seq_length          95092 non-null  int64 \n",
      " 2   sequence            95092 non-null  object\n",
      " 3   pep_ID              95092 non-null  object\n",
      " 4   seqtype             95092 non-null  object\n",
      " 5   version             95092 non-null  object\n",
      " 6   chromosome          95092 non-null  object\n",
      " 7   start               95092 non-null  object\n",
      " 8   end                 95092 non-null  object\n",
      " 9   strand              95092 non-null  object\n",
      " 10  gene_ID             95092 non-null  object\n",
      " 11  transcript_ID       95092 non-null  object\n",
      " 12  gene_biotype        95092 non-null  object\n",
      " 13  transcript_biotype  95092 non-null  object\n",
      " 14  gene_symbol         95092 non-null  object\n",
      " 15  description         95092 non-null  object\n",
      " 16  source              95092 non-null  object\n",
      " 17  accession           95092 non-null  object\n",
      "dtypes: int64(1), object(17)\n",
      "memory usage: 13.8+ MB\n"
     ]
    }
   ],
   "source": [
    "df = df[(df['transcript_biotype'] == \"protein_coding\")]\n",
    "df = df[(df['gene_biotype'] == \"protein_coding\")]\n",
    "df = df.drop('annotation', 1)\n",
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3d169545",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 116646 entries, 0 to 116645\n",
      "Data columns (total 3 columns):\n",
      " #   Column             Non-Null Count   Dtype \n",
      "---  ------             --------------   ----- \n",
      " 0   ID                 116646 non-null  object\n",
      " 1   GENCODE_basic      68747 non-null   object\n",
      " 2   Ensembl_Canonical  23477 non-null   object\n",
      "dtypes: object(3)\n",
      "memory usage: 2.7+ MB\n"
     ]
    }
   ],
   "source": [
    "df2 = pd.read_csv(\"pep_canonical.csv\")\n",
    "df2.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5055d6bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.merge(df, df2, how=\"left\", on=\"ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "81a20057",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.dropna(subset=[\"GENCODE_basic\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d3d1395b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 66823 entries, 0 to 95091\n",
      "Data columns (total 20 columns):\n",
      " #   Column              Non-Null Count  Dtype \n",
      "---  ------              --------------  ----- \n",
      " 0   ID                  66823 non-null  object\n",
      " 1   seq_length          66823 non-null  int64 \n",
      " 2   sequence            66823 non-null  object\n",
      " 3   pep_ID              66823 non-null  object\n",
      " 4   seqtype             66823 non-null  object\n",
      " 5   version             66823 non-null  object\n",
      " 6   chromosome          66823 non-null  object\n",
      " 7   start               66823 non-null  object\n",
      " 8   end                 66823 non-null  object\n",
      " 9   strand              66823 non-null  object\n",
      " 10  gene_ID             66823 non-null  object\n",
      " 11  transcript_ID       66823 non-null  object\n",
      " 12  gene_biotype        66823 non-null  object\n",
      " 13  transcript_biotype  66823 non-null  object\n",
      " 14  gene_symbol         66823 non-null  object\n",
      " 15  description         66823 non-null  object\n",
      " 16  source              66823 non-null  object\n",
      " 17  accession           66823 non-null  object\n",
      " 18  GENCODE_basic       66823 non-null  object\n",
      " 19  Ensembl_Canonical   21793 non-null  object\n",
      "dtypes: int64(1), object(19)\n",
      "memory usage: 10.7+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "84b592f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['duplicate']=df.gene_ID.duplicated()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0ad0271f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>seq_length</th>\n",
       "      <th>sequence</th>\n",
       "      <th>pep_ID</th>\n",
       "      <th>seqtype</th>\n",
       "      <th>version</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>transcript_biotype</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>description</th>\n",
       "      <th>source</th>\n",
       "      <th>accession</th>\n",
       "      <th>GENCODE_basic</th>\n",
       "      <th>Ensembl_Canonical</th>\n",
       "      <th>duplicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>95085</th>\n",
       "      <td>ENSP00000362196.1</td>\n",
       "      <td>783</td>\n",
       "      <td>MARLGNCSLTWAALIILLLPGSLEECGHISVSAPIVHLGDPITASC...</td>\n",
       "      <td>ENSP00000362196.1</td>\n",
       "      <td>pep</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1</td>\n",
       "      <td>36466043</td>\n",
       "      <td>36483278</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000373104.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CSF3R</td>\n",
       "      <td>colony stimulating factor 3 receptor</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>HGNC:2439</td>\n",
       "      <td>GENCODE basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95086</th>\n",
       "      <td>ENSP00000355406.4</td>\n",
       "      <td>836</td>\n",
       "      <td>MARLGNCSLTWAALIILLLPGSLEECGHISVSAPIVHLGDPITASC...</td>\n",
       "      <td>ENSP00000355406.4</td>\n",
       "      <td>pep</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1</td>\n",
       "      <td>36466044</td>\n",
       "      <td>36479519</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000361632.8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CSF3R</td>\n",
       "      <td>colony stimulating factor 3 receptor</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>HGNC:2439</td>\n",
       "      <td>GENCODE basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95088</th>\n",
       "      <td>ENSP00000332180.5</td>\n",
       "      <td>783</td>\n",
       "      <td>MARLGNCSLTWAALIILLLPGSLEECGHISVSAPIVHLGDPITASC...</td>\n",
       "      <td>ENSP00000332180.5</td>\n",
       "      <td>pep</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1</td>\n",
       "      <td>36466096</td>\n",
       "      <td>36479519</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000331941.6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CSF3R</td>\n",
       "      <td>colony stimulating factor 3 receptor</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>HGNC:2439</td>\n",
       "      <td>GENCODE basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95090</th>\n",
       "      <td>ENSP00000507013.1</td>\n",
       "      <td>165</td>\n",
       "      <td>MDAPRRDMELLSNSLAAYAHIRANPESFGLYFVLGVCFGLLLTLCL...</td>\n",
       "      <td>ENSP00000507013.1</td>\n",
       "      <td>pep</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1</td>\n",
       "      <td>36322030</td>\n",
       "      <td>36323645</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000490466.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>EVA1B</td>\n",
       "      <td>eva-1 homolog B</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>HGNC:25558</td>\n",
       "      <td>GENCODE basic</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95091</th>\n",
       "      <td>ENSP00000270824.1</td>\n",
       "      <td>165</td>\n",
       "      <td>MDAPRRDMELLSNSLAAYAHIRANPESFGLYFVLGVCFGLLLTLCL...</td>\n",
       "      <td>ENSP00000270824.1</td>\n",
       "      <td>pep</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1</td>\n",
       "      <td>36322031</td>\n",
       "      <td>36324154</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENST00000270824.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>EVA1B</td>\n",
       "      <td>eva-1 homolog B</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>HGNC:25558</td>\n",
       "      <td>GENCODE basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      ID  seq_length  \\\n",
       "95085  ENSP00000362196.1         783   \n",
       "95086  ENSP00000355406.4         836   \n",
       "95088  ENSP00000332180.5         783   \n",
       "95090  ENSP00000507013.1         165   \n",
       "95091  ENSP00000270824.1         165   \n",
       "\n",
       "                                                sequence             pep_ID  \\\n",
       "95085  MARLGNCSLTWAALIILLLPGSLEECGHISVSAPIVHLGDPITASC...  ENSP00000362196.1   \n",
       "95086  MARLGNCSLTWAALIILLLPGSLEECGHISVSAPIVHLGDPITASC...  ENSP00000355406.4   \n",
       "95088  MARLGNCSLTWAALIILLLPGSLEECGHISVSAPIVHLGDPITASC...  ENSP00000332180.5   \n",
       "95090  MDAPRRDMELLSNSLAAYAHIRANPESFGLYFVLGVCFGLLLTLCL...  ENSP00000507013.1   \n",
       "95091  MDAPRRDMELLSNSLAAYAHIRANPESFGLYFVLGVCFGLLLTLCL...  ENSP00000270824.1   \n",
       "\n",
       "      seqtype version chromosome     start       end strand  ...  \\\n",
       "95085     pep  GRCh38          1  36466043  36483278     -1  ...   \n",
       "95086     pep  GRCh38          1  36466044  36479519     -1  ...   \n",
       "95088     pep  GRCh38          1  36466096  36479519     -1  ...   \n",
       "95090     pep  GRCh38          1  36322030  36323645     -1  ...   \n",
       "95091     pep  GRCh38          1  36322031  36324154     -1  ...   \n",
       "\n",
       "           transcript_ID    gene_biotype transcript_biotype gene_symbol  \\\n",
       "95085  ENST00000373104.5  protein_coding     protein_coding       CSF3R   \n",
       "95086  ENST00000361632.8  protein_coding     protein_coding       CSF3R   \n",
       "95088  ENST00000331941.6  protein_coding     protein_coding       CSF3R   \n",
       "95090  ENST00000490466.2  protein_coding     protein_coding       EVA1B   \n",
       "95091  ENST00000270824.1  protein_coding     protein_coding       EVA1B   \n",
       "\n",
       "                                description       source   accession  \\\n",
       "95085  colony stimulating factor 3 receptor  HGNC Symbol   HGNC:2439   \n",
       "95086  colony stimulating factor 3 receptor  HGNC Symbol   HGNC:2439   \n",
       "95088  colony stimulating factor 3 receptor  HGNC Symbol   HGNC:2439   \n",
       "95090                       eva-1 homolog B  HGNC Symbol  HGNC:25558   \n",
       "95091                       eva-1 homolog B  HGNC Symbol  HGNC:25558   \n",
       "\n",
       "       GENCODE_basic Ensembl_Canonical duplicate  \n",
       "95085  GENCODE basic               NaN      True  \n",
       "95086  GENCODE basic               NaN      True  \n",
       "95088  GENCODE basic               NaN      True  \n",
       "95090  GENCODE basic              True     False  \n",
       "95091  GENCODE basic               NaN      True  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9f34e32f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True    21793\n",
       "Name: Ensembl_Canonical, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.Ensembl_Canonical.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3ece7cd3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True     44912\n",
       "False    21911\n",
       "Name: duplicate, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.duplicate.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2f3c61d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfC = df[(df['Ensembl_Canonical'] == True)]\n",
    "dfD = df[(df['Ensembl_Canonical'] != True)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5053c468",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 21793 entries, 0 to 95090\n",
      "Data columns (total 21 columns):\n",
      " #   Column              Non-Null Count  Dtype \n",
      "---  ------              --------------  ----- \n",
      " 0   ID                  21793 non-null  object\n",
      " 1   seq_length          21793 non-null  int64 \n",
      " 2   sequence            21793 non-null  object\n",
      " 3   pep_ID              21793 non-null  object\n",
      " 4   seqtype             21793 non-null  object\n",
      " 5   version             21793 non-null  object\n",
      " 6   chromosome          21793 non-null  object\n",
      " 7   start               21793 non-null  object\n",
      " 8   end                 21793 non-null  object\n",
      " 9   strand              21793 non-null  object\n",
      " 10  gene_ID             21793 non-null  object\n",
      " 11  transcript_ID       21793 non-null  object\n",
      " 12  gene_biotype        21793 non-null  object\n",
      " 13  transcript_biotype  21793 non-null  object\n",
      " 14  gene_symbol         21793 non-null  object\n",
      " 15  description         21793 non-null  object\n",
      " 16  source              21793 non-null  object\n",
      " 17  accession           21793 non-null  object\n",
      " 18  GENCODE_basic       21793 non-null  object\n",
      " 19  Ensembl_Canonical   21793 non-null  object\n",
      " 20  duplicate           21793 non-null  bool  \n",
      "dtypes: bool(1), int64(1), object(19)\n",
      "memory usage: 3.5+ MB\n"
     ]
    }
   ],
   "source": [
    "dfC.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "46bb9c24",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 45030 entries, 14 to 95091\n",
      "Data columns (total 21 columns):\n",
      " #   Column              Non-Null Count  Dtype \n",
      "---  ------              --------------  ----- \n",
      " 0   ID                  45030 non-null  object\n",
      " 1   seq_length          45030 non-null  int64 \n",
      " 2   sequence            45030 non-null  object\n",
      " 3   pep_ID              45030 non-null  object\n",
      " 4   seqtype             45030 non-null  object\n",
      " 5   version             45030 non-null  object\n",
      " 6   chromosome          45030 non-null  object\n",
      " 7   start               45030 non-null  object\n",
      " 8   end                 45030 non-null  object\n",
      " 9   strand              45030 non-null  object\n",
      " 10  gene_ID             45030 non-null  object\n",
      " 11  transcript_ID       45030 non-null  object\n",
      " 12  gene_biotype        45030 non-null  object\n",
      " 13  transcript_biotype  45030 non-null  object\n",
      " 14  gene_symbol         45030 non-null  object\n",
      " 15  description         45030 non-null  object\n",
      " 16  source              45030 non-null  object\n",
      " 17  accession           45030 non-null  object\n",
      " 18  GENCODE_basic       45030 non-null  object\n",
      " 19  Ensembl_Canonical   0 non-null      object\n",
      " 20  duplicate           45030 non-null  bool  \n",
      "dtypes: bool(1), int64(1), object(19)\n",
      "memory usage: 7.3+ MB\n"
     ]
    }
   ],
   "source": [
    "dfD.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "85076297",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True     37847\n",
       "False     7183\n",
       "Name: duplicate, dtype: int64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfD.duplicate.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "5ed60aee",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfD = dfD[(dfD['duplicate'] == False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2349f533",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfcomplete = pd.concat([dfC, dfD])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "2d25e107",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 28976 entries, 0 to 95071\n",
      "Data columns (total 21 columns):\n",
      " #   Column              Non-Null Count  Dtype \n",
      "---  ------              --------------  ----- \n",
      " 0   ID                  28976 non-null  object\n",
      " 1   seq_length          28976 non-null  int64 \n",
      " 2   sequence            28976 non-null  object\n",
      " 3   pep_ID              28976 non-null  object\n",
      " 4   seqtype             28976 non-null  object\n",
      " 5   version             28976 non-null  object\n",
      " 6   chromosome          28976 non-null  object\n",
      " 7   start               28976 non-null  object\n",
      " 8   end                 28976 non-null  object\n",
      " 9   strand              28976 non-null  object\n",
      " 10  gene_ID             28976 non-null  object\n",
      " 11  transcript_ID       28976 non-null  object\n",
      " 12  gene_biotype        28976 non-null  object\n",
      " 13  transcript_biotype  28976 non-null  object\n",
      " 14  gene_symbol         28976 non-null  object\n",
      " 15  description         28976 non-null  object\n",
      " 16  source              28976 non-null  object\n",
      " 17  accession           28976 non-null  object\n",
      " 18  GENCODE_basic       28976 non-null  object\n",
      " 19  Ensembl_Canonical   21793 non-null  object\n",
      " 20  duplicate           28976 non-null  bool  \n",
      "dtypes: bool(1), int64(1), object(19)\n",
      "memory usage: 4.7+ MB\n"
     ]
    }
   ],
   "source": [
    "dfcomplete.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b527aa84",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfcomplete['duplicate']=dfcomplete.gene_ID.duplicated()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "baea5363",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    21911\n",
       "True      7065\n",
       "Name: duplicate, dtype: int64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfcomplete.duplicate.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e4cefd78",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 21911 entries, 0 to 94197\n",
      "Data columns (total 21 columns):\n",
      " #   Column              Non-Null Count  Dtype \n",
      "---  ------              --------------  ----- \n",
      " 0   ID                  21911 non-null  object\n",
      " 1   seq_length          21911 non-null  int64 \n",
      " 2   sequence            21911 non-null  object\n",
      " 3   pep_ID              21911 non-null  object\n",
      " 4   seqtype             21911 non-null  object\n",
      " 5   version             21911 non-null  object\n",
      " 6   chromosome          21911 non-null  object\n",
      " 7   start               21911 non-null  object\n",
      " 8   end                 21911 non-null  object\n",
      " 9   strand              21911 non-null  object\n",
      " 10  gene_ID             21911 non-null  object\n",
      " 11  transcript_ID       21911 non-null  object\n",
      " 12  gene_biotype        21911 non-null  object\n",
      " 13  transcript_biotype  21911 non-null  object\n",
      " 14  gene_symbol         21911 non-null  object\n",
      " 15  description         21911 non-null  object\n",
      " 16  source              21911 non-null  object\n",
      " 17  accession           21911 non-null  object\n",
      " 18  GENCODE_basic       21911 non-null  object\n",
      " 19  Ensembl_Canonical   21793 non-null  object\n",
      " 20  duplicate           21911 non-null  bool  \n",
      "dtypes: bool(1), int64(1), object(19)\n",
      "memory usage: 3.5+ MB\n"
     ]
    }
   ],
   "source": [
    "dfcomplete = dfcomplete[(dfcomplete['Ensembl_Canonical'] == True) | (dfcomplete['duplicate'] == False)]\n",
    "dfcomplete.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "9e80bdb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrlist = [\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"8\", \"9\", \"10\", \"11\", \"12\", \"13\", \"14\", \"15\", \"16\", \"17\", \"18\", \"19\", \"20\", \"21\", \"22\", \"X\", \"Y\" ]\n",
    "dfcomplete = dfcomplete[dfcomplete[\"chromosome\"].isin(chrlist)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "c154905c",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfcomplete.to_csv('HG30_pep_annotation.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "0bd26b0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_id = []\n",
    "gene_name = []\n",
    "seq = []\n",
    "\n",
    "output_path = 'HG38_pep_unique.fasta'\n",
    "output_file = open(output_path,'w')\n",
    "\n",
    "for i, row in dfcomplete.iterrows():\n",
    "    seq_id = dfcomplete['ID'][i]\n",
    "    gene_name = dfcomplete['gene_symbol'][i]\n",
    "    seq = dfcomplete['sequence'][i]\n",
    "    identifier_line = \">\" + seq_id + \"\\t\" + gene_name + \"\\n\"\n",
    "    output_file.write(identifier_line)\n",
    "    sequence_line = seq + \"\\n\"\n",
    "    output_file.write(sequence_line)\n",
    "    \n",
    "#Close the file when we're done\n",
    "output_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c8a7e21",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
