library(GenomicRanges)
library(SummarizedExperiment)
library(JASPAR2022)
library(TFBSTools)
library(BSgenome.Hsapiens.UCSC.hg38)
library(monaLisa)
library(ComplexHeatmap)
library(circlize)


WT.unique <- rtracklayer::import(con = "ZIP13K2_NGN2_WT.unique.bed", format = "bed")
#AroLITE.unique <- rtracklayer::import(con = "ZIP13K2_NGN2_AroLITE.unique.bed", format = "bed")
AroPERFECT.unique <- rtracklayer::import(con = "ZIP13K2_NGN2_AroPERFECT.unique.bed", format = "bed")
NGN2_common <- rtracklayer::import(con = "ZIP13K2_NGN2_common.All.bed", format = "bed")


hist(WT.unique$score, 100, col = "gray", main = "",
     xlab = "Chipseq peaks score", ylab = "Number of peaks")
#hist(AroLITE.unique$score, 100, col = "gray", main = "",
#     xlab = "Chipseq peaks score", ylab = "Number of peaks")
hist(AroPERFECT.unique$score, 100, col = "gray", main = "",
     xlab = "Chipseq peaks score", ylab = "Number of peaks")
hist(NGN2_common$score, 100, col = "gray", main = "",
     xlab = "Chipseq peaks score", ylab = "Number of peaks")


pwms <- getMatrixSet(JASPAR2022,
                     opts = list(matrixtype = "PWM",
                                 tax_group = "vertebrates"))


MergedSet <- c(WT.unique, AroPERFECT.unique, NGN2_common)


# get TFBS on given GRanges (peaks)
# suppress warnings generated by matchPWM due to the presence of Ns 
# in the sequences
peakMerged <- getSeq(BSgenome.Hsapiens.UCSC.hg38, MergedSet)

motif_list <- c("MA0048.1", "MA0607.2", "MA0623.2", "MA0665.1", "MA0668.1", "MA0669.1", "MA0678.1", "MA0817.1", "MA0818.1", "MA0820.1", "MA0826.1", "MA0827.1", "MA1109.1", "MA1123.1", "MA1467.1", "MA1468.1", "MA1472.1", "MA1485.1", "MA1524.1", "MA1529.1", "MA1568.1", "MA1635.1", "MA1638.1", "MA1642.1", "MA0819.1", "MA0499.2", "MA1099.2", "MA0616.2", "MA0058.1")

pfms <- getMatrixByID(JASPAR2022, c("MA0048.1", "MA0607.2", "MA0623.2", "MA0665.1", "MA0668.1", "MA0669.1", "MA0678.1", "MA0817.1", "MA0818.1", "MA0820.1", "MA0826.1", "MA0827.1", "MA1109.1", "MA1123.1", "MA1467.1", "MA1468.1", "MA1472.1", "MA1485.1", "MA1524.1", "MA1529.1", "MA1568.1", "MA1635.1", "MA1638.1", "MA1642.1", "MA0819.1", "MA0499.2", "MA1099.2", "MA0616.2", "MA0058.1"))
pwms <- toPWM(pfms)
```


# define binning vector
bins <- rep(c("WT.unique", "AroPERFECT.unique", "NGN2_common"), c(length(WT.unique), length(AroPERFECT.unique), length(NGN2_common)))
bins <- factor(bins)
table(bins)


se <- calcBinnedMotifEnrR(seqs = peakMerged, bins = bins,
                          background = "genome",
                          genome = BSgenome.Hsapiens.UCSC.hg38,
                          genome.regions = NULL, # sample from full genome
                          genome.oversample = 2,
                          BPPARAM = BiocParallel::SerialParam(RNGseed = 42),
                          pwmL = pwms)
se

se2 <- calcBinnedMotifEnrR(seqs = peakMerged, bins = bins,
                           background = "allBins",
                           genome = BSgenome.Hsapiens.UCSC.hg38,
                           genome.regions = NULL, # sample from full genome
                           genome.oversample = 2,
                           BPPARAM = BiocParallel::SerialParam(RNGseed = 42),
                           pwmL = pwms)
se2

se3 <- calcBinnedMotifEnrR(seqs = peakMerged, bins = bins,
                           background = "otherBins",
                           genome = BSgenome.Hsapiens.UCSC.hg38,
                           genome.regions = NULL, # sample from full genome
                           genome.oversample = 2,
                           BPPARAM = BiocParallel::SerialParam(RNGseed = 42),
                           pwmL = pwms)
se3


SimMatSel <- motifSimilarity(rowData(se)$motif.pfm)
hcl <- hclust(as.dist(1 - SimMatSel), method = "average")


# plot
plotMotifHeatmaps(x = se, which.plots = c("log2enr", "negLog10Padj"), 
                  width = 1.0, cluster = hcl,  maxEnr = 4, maxSig = 20, 
                  show_motif_GC = F, show_seqlogo = T,show_dendrogram = TRUE, doPlot = T)

plotMotifHeatmaps(x = se2, which.plots = c("log2enr", "negLog10Padj"), 
                  width = 1.0, cluster = hcl,  maxEnr = 4, maxSig = 20, 
                  show_motif_GC = F, show_seqlogo = T,show_dendrogram = TRUE, doPlot = T)

plotMotifHeatmaps(x = se3, which.plots = c("log2enr", "negLog10Padj"), 
                  width = 1.0, cluster = hcl,  maxEnr = 4, maxSig = 20, 
                  show_motif_GC = F, show_seqlogo = T,show_dendrogram = TRUE, doPlot = T)

